app/services/onetime/backfill_payment_option_installment_snapshots.rb

```diff
@@ -9,18 +9,48 @@ def self.perform
         next unless payment_option.installment_plan.present?
         next unless payment_option.subscription&.original_purchase.present?
 
-        original_purchase = payment_option.subscription.original_purchase
-        total_price = original_purchase.total_price_before_installments || original_purchase.price_cents
+        subscription = payment_option.subscription
+        installment_plan = payment_option.installment_plan
+        
+        total_price = calculate_total_price_from_history(subscription, installment_plan)
 
         InstallmentPlanSnapshot.create!(
           payment_option: payment_option,
-          number_of_installments: payment_option.installment_plan.number_of_installments,
-          recurrence: payment_option.installment_plan.recurrence,
+          number_of_installments: installment_plan.number_of_installments,
+          recurrence: installment_plan.recurrence,
           total_price_cents: total_price
         )
       rescue StandardError => e
         Rails.logger.error("Failed to backfill PaymentOption #{payment_option.id}: #{e.message}")
       end
     end
+
+    def self.calculate_total_price_from_history(subscription, installment_plan)
+      original_purchase = subscription.original_purchase
+      
+      all_installment_purchases = subscription.purchases
+        .successful
+        .is_installment_payment
+        .order(:created_at)
+      
+      if all_installment_purchases.count > 1
+        total_paid = all_installment_purchases.sum(:price_cents)
+        completed_installments = all_installment_purchases.count
+        expected_installments = installment_plan.number_of_installments
```

Comment:
Naming matters. When I read `completed_installments`, I expect a collection of Purchase records, but this is actually a count. You should name your variables based on what they actually return to avoid misleading other contributors or yourself in the future. Example: `completed_installments_count`, or `number_of_completed_installments`.

Same with expected_installments`